src/main/java/com/cloudbees/jenkins/support/threaddump/HighLoadChecker.java

```diff
@@ -0,0 +1,103 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.jenkins.support.threaddump;
+
+import com.cloudbees.jenkins.support.impl.ThreadDumps;
+import com.cloudbees.jenkins.support.timer.FileListCap;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricSet;
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.model.PeriodicWork;
+import jenkins.metrics.api.MetricProvider;
+import jenkins.metrics.impl.VMMetricProviderImpl;
+import jenkins.model.Jenkins;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * PeriodicWork to check when there is a high load in the instance.
+ *
+ * Only checking high CPU usage for the moment, but it can be used
+ * to generate thread dumps in high heap memory consumption.
+ */
+@Extension
+public class HighLoadChecker extends PeriodicWork {
+
+    /**
+     * Recurrence period to check high load consumption and take the corresponded actions
+     * like generating a thread dump
+     */
+    public static final int RECURRENCE_PERIOD_SEC =
+            Integer.getInteger(HighLoadChecker.class.getName()+ ".RECURRENCE_PERIOD_SEC", 900);
+
+    /**
+     * This is the CPU usage threshold
+     */
+    public static final Double CPU_USAGE_THRESHOLD =
+            new Double(System.getProperty(HighLoadChecker.class.getName() + ".CPU_USAGE_THRESHOLD", "0.95"));
+
+    /**
+     * Limit the number of thread dumps to retain on high cpu
+     */
+    public static final int HIGH_CPU_THEAD_DUMPS_TO_RETAIN =
+            Integer.getInteger(HighLoadChecker.class.getName()+ ".HIGH_CPU_THEAD_DUMPS_TO_RETAIN", 10);
+
+    /**
+     * Thread dumps generated on high CPU load are stored in $JENKINS_HOME/high-load/cpu
+     **/
+    final FileListCap logs = new FileListCap(new File(Jenkins.getInstanceOrNull().getRootDir(),"high-load/cpu"), HIGH_CPU_THEAD_DUMPS_TO_RETAIN);
+
+    final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+    {
+        format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return TimeUnit.SECONDS.toMillis(RECURRENCE_PERIOD_SEC);
+    }
+
+    @Override
+    protected void doRun() throws Exception {
+        VMMetricProviderImpl vmMetricProvider = ExtensionList.lookup(MetricProvider.class).get(VMMetricProviderImpl.class);
+        MetricSet vmProviderMetricSet = vmMetricProvider.getMetricSet();
+        Gauge cpu = (Gauge) vmProviderMetricSet.getMetrics().get("vm.cpu.load");
+        Double cpuLoad = new Double(cpu.getValue().toString());
+
+        if (Double.compare(CPU_USAGE_THRESHOLD, cpuLoad) < 0) {
+            File threadDumpFile = logs.file(format.format(new Date()) + ".txt");
+            logs.add(threadDumpFile);
```

Comment:
Maybe this should be at the end, so that we don't save files that cannot be correctly generated?